c-dependencies/js-compute-runtime/js-compute-builtins.cpp

```diff
@@ -694,11 +694,244 @@ bool parse_body(JSContext *cx, HandleObject self, UniqueChars buf, size_t len) {
+  RootedValue done_val(cx);
+  RootedValue value(cx);
+#ifdef DEBUG
+  bool hasValue;
+  if (!JS_HasProperty(cx, chunk_obj, "value", &hasValue)) {
+    return false;
+  }
+  MOZ_ASSERT(hasValue);
+#endif
+  if (!JS_GetProperty(cx, chunk_obj, "value", &value)) {
+    return false;
+  }
+#ifdef DEBUG
+  bool hasDone;
+  if (!JS_HasProperty(cx, chunk_obj, "done", &hasDone)) {
+    return false;
+  }
+  MOZ_ASSERT(hasDone);
+#endif
+  if (!JS_GetProperty(cx, chunk_obj, "done", &done_val)) {
+    return false;
+  }
+  MOZ_ASSERT(done_val.isBoolean());
+  if (done_val.toBoolean()) {
+    // We finished reading the stream
+    // Now we need to iterate/reduce `contents` JS Array into UniqueChars
+    uint32_t contentsLength;
+    if (!JS::GetArrayLength(cx, contents, &contentsLength)) {
+      return false;
+    }
+    std::vector<uint8_t> allBytes;
+    for (uint32_t index = 0; index < contentsLength; index++) {
+      RootedValue val(cx);
+      if (!JS_GetElement(cx, contents, index, &val)) {
+        return false;
+      }
+      {
+        JS::AutoCheckCannotGC nogc;
+        MOZ_ASSERT(val.isObject());
+        JSObject *array = &val.toObject();
+        MOZ_ASSERT(JS_IsUint8Array(array));
+        bool is_shared;
+        uint8_t *bytes = JS_GetUint8ArrayData(array, &is_shared, nogc);
+        size_t length = JS_GetTypedArrayByteLength(array);
+        allBytes.insert(allBytes.end(), bytes, bytes + length);
+      }
+    }
+#ifdef DEBUG
+    bool foundBodyParser;
+    if (!JS_HasElement(cx, catch_handler, 2, &foundBodyParser)) {
+      return false;
+    }
+    MOZ_ASSERT(foundBodyParser);
+#endif
+    // Now we can call parse_body on the result
+    RootedValue body_parser(cx);
+    if (!JS_GetElement(cx, catch_handler, 2, &body_parser)) {
+      return false;
+    }
+    auto parse_body = (ParseBodyCB *)body_parser.toPrivate();
+    size_t bytes_read = allBytes.size();
+    UniqueChars buf(reinterpret_cast<char *>(allBytes.data()));
+    return parse_body(cx, self, std::move(buf), bytes_read);
+  }
+
+  RootedValue val(cx);
+  if (!JS_GetProperty(cx, chunk_obj, "value", &val)) {
+    return false;
+  }
+
+  // The read operation can return anything since this stream comes from the guest
+  // If it is not a UInt8Array -- reject with a TypeError
+  if (!val.isObject() || !JS_IsUint8Array(&val.toObject())) {
+    JS_ReportErrorNumberASCII(cx, GetErrorMessage, nullptr, JSMSG_RESPONSE_VALUE_NOT_UINT8ARRAY);
+    RootedObject result_promise(cx);
+    result_promise = &JS::GetReservedSlot(self, Slots::BodyAllPromise).toObject();
+    JS::SetReservedSlot(self, Slots::BodyAllPromise, JS::UndefinedValue());
+    return RejectPromiseWithPendingError(cx, result_promise);
+  }
+
+  {
+    uint32_t contentsLength;
+    if (!JS::GetArrayLength(cx, contents, &contentsLength)) {
+      return false;
+    }
+    std::string index = std::to_string(contentsLength);
+    if (!JS_SetProperty(cx, contents, index.c_str(), val)) {
+      return false;
+    }
```

Comment:
Will `JS_SetProperty` parse the number out of the string when setting the property?

Reply:
`JS_SetProperty` won't parse the number out of the string - we can actually avoid the conversion to a string here by using `JS_SetElement` -- I've now moved to using that instead 👍